msamp/megatron/layers.py

```diff
@@ -50,19 +58,32 @@ def forward(ctx, input, weight, bias, gradient_accumulation_fusion, async_grad_a
 
         old_meta_group = input_meta.group
         input_meta.group = tp_group
-        input_fp8 = input.cast(Dtypes.kfloat8_e4m3, meta=input_meta, sync=sequence_parallel)
+        if USE_A_SIMU_FP4:
+            fp4_input_in_float = FP4_QUANT.quantize_simu_fp4_in_bf16(input.bfloat16(), format='e2m1', nan_existed=False, token_wise=True, outlier_clip=True, clip_threshold=0.99)
+            input_fp8 = fp4_input_in_float.cast(Dtypes.kfloat8_e4m3, meta=input_meta, sync=sequence_parallel)
```

Comment:
Why not we just write something like
```
if USE_A_SIMU_FP4:
    input = FP4_QUANT.quantize_simu_fp4_in_bf16(input.bfloat16(), format='e2m1', nan_existed=False, token_wise=True, outlier_clip=True, clip_threshold=0.99)
input_fp8 = input.cast(Dtypes.kfloat8_e4m3, meta=input_meta, sync=sequence_parallel)
```
since we're doing the same casting in the end.